File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/groupsv2/DecryptedGroupUtilTest.java
package org.whispersystems.signalservice.api.groupsv2;

import org.junit.Test;
import org.signal.storageservice.protos.groups.local.DecryptedGroupChange;
import org.signal.storageservice.protos.groups.local.DecryptedPendingMember;
import org.signal.storageservice.protos.groups.local.DecryptedPendingMemberRemoval;
import org.whispersystems.signalservice.api.push.ServiceId;
import org.whispersystems.signalservice.api.push.ServiceId.ACI;
import org.whispersystems.signalservice.internal.util.Util;

import java.util.List;
import java.util.UUID;

import okio.ByteString;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;
import static java.util.Arrays.asList;

public final class DecryptedGroupUtilTest {

  @Test
  public void can_extract_editor_uuid_from_decrypted_group_change() {
    ACI        aci                   = ACI.from(UUID.randomUUID());
    ByteString editor                = aci.toByteString();
    DecryptedGroupChange groupChange = new DecryptedGroupChange.Builder()
                                                           .editorServiceIdBytes(editor)
                                                           .build();

    ServiceId parsed = DecryptedGroupUtil.editorServiceId(groupChange).get();

    assertEquals(aci, parsed);
  }

  @Test
  public void can_extract_uuid_from_decrypted_pending_member() {
    ACI                    aci             = ACI.from(UUID.randomUUID());
    DecryptedPendingMember decryptedMember = new DecryptedPendingMember.Builder()
                                                                   .serviceIdBytes(aci.toByteString())
                                                                   .build();

    ServiceId parsed = ServiceId.parseOrNull(decryptedMember.serviceIdBytes);

    assertEquals(aci, parsed);
  }

  @Test
  public void can_extract_uuid_from_bad_decrypted_pending_member() {
    DecryptedPendingMember decryptedMember = new DecryptedPendingMember.Builder()
                                                                   .serviceIdBytes(ByteString.of(Util.getSecretBytes(18)))
                                                                   .build();

    ServiceId parsed = ServiceId.parseOrNull(decryptedMember.serviceIdBytes);

    assertNull(parsed);
  }

  @Test
  public void can_extract_uuids_for_all_pending_including_bad_entries() {
    ACI                    aci1             = ACI.from(UUID.randomUUID());
    ACI                    aci2             = ACI.from(UUID.randomUUID());
    DecryptedPendingMember decryptedMember1 = new DecryptedPendingMember.Builder()
                                                                    .serviceIdBytes(aci1.toByteString())
                                                                    .build();
    DecryptedPendingMember decryptedMember2 = new DecryptedPendingMember.Builder()
                                                                    .serviceIdBytes(aci2.toByteString())
                                                                    .build();
    DecryptedPendingMember decryptedMember3 = new DecryptedPendingMember.Builder()
                                                                    .serviceIdBytes(ByteString.of(Util.getSecretBytes(18)))
                                                                    .build();

    DecryptedGroupChange groupChange = new DecryptedGroupChange.Builder()
                                                               .newPendingMembers(asList(decryptedMember1, decryptedMember2, decryptedMember3))
                                                               .build();

    List<ServiceId> pendingUuids = DecryptedGroupUtil.pendingToServiceIdList(groupChange.newPendingMembers);

    assertThat(pendingUuids, is(asList(aci1, aci2, ACI.UNKNOWN)));
  }

  @Test
  public void can_extract_uuids_for_all_deleted_pending_excluding_bad_entries() {
    ACI                           aci1             = ACI.from(UUID.randomUUID());
    ACI                           aci2             = ACI.from(UUID.randomUUID());
    DecryptedPendingMemberRemoval decryptedMember1 = new DecryptedPendingMemberRemoval.Builder()
                                                                    .serviceIdBytes(aci1.toByteString())
                                                                    .build();
    DecryptedPendingMemberRemoval decryptedMember2 = new DecryptedPendingMemberRemoval.Builder()
                                                                    .serviceIdBytes(aci2.toByteString())
                                                                    .build();
    DecryptedPendingMemberRemoval decryptedMember3 = new DecryptedPendingMemberRemoval.Builder()
                                                                    .serviceIdBytes(ByteString.of(Util.getSecretBytes(18)))
                                                                    .build();

    DecryptedGroupChange groupChange = new DecryptedGroupChange.Builder()
                                                               .deletePendingMembers(asList(decryptedMember1, decryptedMember2, decryptedMember3))
                                                               .build();

    List<ServiceId> removedUuids = DecryptedGroupUtil.removedPendingMembersServiceIdList(groupChange);

    assertThat(removedUuids, is(asList(aci1, aci2)));
  }

  @Test
  public void can_extract_uuids_for_all_deleted_members_excluding_bad_entries() {
    ACI                  aci1        = ACI.from(UUID.randomUUID());
    ACI                  aci2        = ACI.from(UUID.randomUUID());
    DecryptedGroupChange groupChange = new DecryptedGroupChange.Builder()
                                                               .deleteMembers(asList(aci1.toByteString(), aci2.toByteString(), ByteString.of(Util.getSecretBytes(18))))
                                                               .build();

    List<ServiceId> removedServiceIds = DecryptedGroupUtil.removedMembersServiceIdList(groupChange);

    assertThat(removedServiceIds, is(asList(aci1, aci2)));
  }
}

File: libsignal-service/src/test/java/org/whispersystems/signalservice/api/groupsv2/DecryptedGroupUtilTest.kt
package org.whispersystems.signalservice.api.groupsv2

import assertk.assertThat
import assertk.assertions.containsExactly
import assertk.assertions.isEqualTo
import assertk.assertions.isNull
import okio.ByteString
import org.junit.Test
import org.signal.storageservice.protos.groups.local.DecryptedGroupChange
import org.signal.storageservice.protos.groups.local.DecryptedPendingMember
import org.signal.storageservice.protos.groups.local.DecryptedPendingMemberRemoval
import org.whispersystems.signalservice.api.push.ServiceId
import org.whispersystems.signalservice.internal.util.Util
import java.util.UUID

class DecryptedGroupUtilTest {
  @Test
  fun can_extract_editor_uuid_from_decrypted_group_change() {
    val aci = randomACI()
    val editor = aci.toByteString()
    val groupChange = DecryptedGroupChange.Builder()
      .editorServiceIdBytes(editor)
      .build()

    val parsed = DecryptedGroupUtil.editorServiceId(groupChange).get()

    assertThat(parsed).isEqualTo(aci)
  }

  @Test
  fun can_extract_uuid_from_decrypted_pending_member() {
    val aci = randomACI()
    val decryptedMember = DecryptedPendingMember.Builder()
      .serviceIdBytes(aci.toByteString())
      .build()

    val parsed = ServiceId.parseOrNull(decryptedMember.serviceIdBytes)

    assertThat(parsed).isEqualTo(aci)
  }

  @Test
  fun can_extract_uuid_from_bad_decrypted_pending_member() {
    val decryptedMember = DecryptedPendingMember.Builder()
      .serviceIdBytes(ByteString.of(*Util.getSecretBytes(18)))
      .build()

    val parsed = ServiceId.parseOrNull(decryptedMember.serviceIdBytes)

    assertThat(parsed).isNull()
  }

  @Test
  fun can_extract_uuids_for_all_pending_including_bad_entries() {
    val aci1 = randomACI()
    val aci2 = randomACI()
    val decryptedMember1 = DecryptedPendingMember.Builder()
      .serviceIdBytes(aci1.toByteString())
      .build()
    val decryptedMember2 = DecryptedPendingMember.Builder()
      .serviceIdBytes(aci2.toByteString())
      .build()
    val decryptedMember3 = DecryptedPendingMember.Builder()
      .serviceIdBytes(ByteString.of(*Util.getSecretBytes(18)))
      .build()

    val groupChange = DecryptedGroupChange.Builder()
      .newPendingMembers(listOf(decryptedMember1, decryptedMember2, decryptedMember3))
      .build()

    val pendingUuids = DecryptedGroupUtil.pendingToServiceIdList(groupChange.newPendingMembers)

    assertThat(pendingUuids).containsExactly(aci1, aci2, ServiceId.ACI.UNKNOWN)
  }

  @Test
  fun can_extract_uuids_for_all_deleted_pending_excluding_bad_entries() {
    val aci1 = randomACI()
    val aci2 = randomACI()
    val decryptedMember1 = DecryptedPendingMemberRemoval.Builder()
      .serviceIdBytes(aci1.toByteString())
      .build()
    val decryptedMember2 = DecryptedPendingMemberRemoval.Builder()
      .serviceIdBytes(aci2.toByteString())
      .build()
    val decryptedMember3 = DecryptedPendingMemberRemoval.Builder()
      .serviceIdBytes(ByteString.of(*Util.getSecretBytes(18)))
      .build()

    val groupChange = DecryptedGroupChange.Builder()
      .deletePendingMembers(listOf(decryptedMember1, decryptedMember2, decryptedMember3))
      .build()

    val removedUuids = DecryptedGroupUtil.removedPendingMembersServiceIdList(groupChange)

    assertThat(removedUuids).containsExactly(aci1, aci2)
  }

  @Test
  fun can_extract_uuids_for_all_deleted_members_excluding_bad_entries() {
    val aci1 = randomACI()
    val aci2 = randomACI()
    val groupChange = DecryptedGroupChange.Builder()
      .deleteMembers(listOf(aci1.toByteString(), aci2.toByteString(), ByteString.of(*Util.getSecretBytes(18))))
      .build()

    val removedServiceIds = DecryptedGroupUtil.removedMembersServiceIdList(groupChange)

    assertThat(removedServiceIds).containsExactly(aci1, aci2)
  }

  private fun randomACI() = ServiceId.ACI.from(UUID.randomUUID())
}
